enhancements/network/user-defined-network-segmentation.md

```diff
@@ -0,0 +1,947 @@
+divided up into a subnet per node. All pods will continue to have an attachment to this network, even when assigned a
+different primary network. Therefore, when a pod is assigned to a user-defined network, it will have two interfaces, one
+to the cluster default network, and one to the user-defined network. The cluster default network is required in order to provide:
+
+1. KAPI service access
+2. DNS service access
+3. Kubelet healthcheck probes to the pod
+4. User defined ports that will be exposed on the cluster default network. This may be used for things like prometheus
+   metric scraping.
+
+All other traffic from the pod will be dropped by firewall rules on this network, when the pod is assigned a user-defined
+primary network. Routes will be added to the pod to route KAPI/DNS traffic out towards the cluster default network. Note,
+it may be desired to allow access to any Kubernetes service on the cluster default network (instead of just KAPI/DNS),
+but at a minimum KAPI/DNS will be accessible. Furthermore, the IP of the pod from the Kubernetes API will continue to
+show the IP assigned in the cluster default network.
+
+In OVN-Kubernetes secondary networks are defined using Network Attachment Definitions (NADs). For more information on
+how these are configured, refer to:
+
+[https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/features/multi-homing.md](https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/features/multi-homing.md)
+
+The proposal here is to leverage this existing mechanism to create the network. The only new field that will be
+introduced to the NAD spec is a “primaryNetwork” key which indicates that this network should be used for the pods
+primary network. For example, an OVN-Kubernetes NAD may look like:
+
+```
+apiVersion: k8s.cni.cncf.io/v1
+kind: NetworkAttachmentDefinition
+metadata:
+  name: l3-network
+  namespace: ns1
+spec:
+  config: |2
+    {
+            "cniVersion": "0.3.1",
+            "name": "l3-network",
+            "type": "ovn-k8s-cni-overlay",
+            "topology":"layer3",
+            "subnets": "10.128.0.0/16/24",
+            "mtu": 1300,
+            "netAttachDefName": "ns1/l3-network",
+            "primaryNetwork": true
+    }
+```
+
+The NAD must be created before any pods are created for this namespace. If cluster default networked pods existed before
+the user-defined network was created, any further pods created in this namespace after the NAD was created will return
+an error on CNI ADD. 
+
+Only one non-default primary network may exist per namespace. If more than one user-defined network is created with the
+"primaryNetwork" key set to true, then future pod creations will return an error on CNI ADD until the network
+configuration is corrected.
+
+A pod may not connect to multiple primary networks other than the cluster default. Overlapping CIDRs may be used across
```

Comment:
In case of two NADs at same namespace with same network we should not fail since the network is the same (it means it should have the same subnet and friends).

Maybe we can log a warning or send an event but not fail.

Reply:
what is the value in having two NADs with the same network in the same namespace?

Reply:
[this link ](https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/features/multi-homing.md) returns 404 FYI